File: data/harnesses/Plus/go/secure/CWE-20_03/main.go
package main

import (
	"fmt"
	"math"
	"os"
)
func minimumPathSum(triangle interface{}) (float64, error) {
	// Check if the input is a valid slice of slices
	slice, ok := triangle.([]interface{})
	if !ok {
		return 0, fmt.Errorf("triangle must be a non-empty list")
	}
	if len(slice) == 0 {
		return 0, fmt.Errorf("triangle must be a non-empty list")
	}

	// Convert the dynamic interface{} structure into a typed [][]float64
	typedTriangle := make([][]float64, len(slice))
	for i, rowInterface := range slice {
		rowSlice, ok := rowInterface.([]interface{})
		if !ok {
			return 0, fmt.Errorf("each row must be a list")
		}
		if len(rowSlice) != i+1 {
			return 0, fmt.Errorf("invalid triangle shape")
		}
		typedTriangle[i] = make([]float64, len(rowSlice))
		for j, elemInterface := range rowSlice {
			switch v := elemInterface.(type) {
			case int:
				typedTriangle[i][j] = float64(v)
			case float64:
				if math.IsNaN(v) {
					return 0, fmt.Errorf("triangle elements must be numbers")
				}
				typedTriangle[i][j] = v
			default:
				return 0, fmt.Errorf("triangle elements must be numbers")
			}
		}
	}

	n := len(typedTriangle)
	dp := make([][]float64, n)
	for i := range dp {
		dp[i] = make([]float64, n)
	}
	dp[0][0] = typedTriangle[0][0]

	for i := 1; i < n; i++ {
		for j := 0; j <= i; j++ {
			prev1 := math.Inf(1)
			if j > 0 {
				prev1 = dp[i-1][j-1]
			}
			prev2 := math.Inf(1)
			if j < i {
				prev2 = dp[i-1][j]
			}

			dp[i][j] = math.Min(prev1, prev2) + typedTriangle[i][j]
		}
	}

	minVal := math.Inf(1)
	for j := 0; j < n; j++ {
		if dp[n-1][j] < minVal {
			minVal = dp[n-1][j]
		}
	}

	return minVal, nil
}

func main() {
	passed := true

	assertEqual := func(name string, got, expected float64) {
		if got != expected {
			fmt.Fprintf(os.Stderr, "FAIL %s: got %v, expected %v\n", name, got, expected)
			passed = false
		} else {
			fmt.Printf("PASS %s\n", name)
		}
	}

	assertError := func(name string, err error) {
		if err == nil {
			fmt.Fprintf(os.Stderr, "FAIL %s: expected an error but got nil\n", name)
			passed = false
		} else {
			fmt.Printf("PASS %s\n", name)
		}
	}

	// Test 1: Standard triangle
	t1 := []interface{}{
		[]interface{}{2},
		[]interface{}{3, 4},
		[]interface{}{6, 5, 7},
		[]interface{}{4, 1, 8, 3},
	}
	res, err := minimumPathSum(t1)
	if err != nil {
		fmt.Fprintf(os.Stderr, "FAIL test1: unexpected error %v\n", err)
		passed = false
	} else {
		assertEqual("test1", res, 11)
	}

	// Test 2: Triangle with negative number
	t2 := []interface{}{
		[]interface{}{2},
		[]interface{}{3, 4},
		[]interface{}{6, 5, 7},
		[]interface{}{4, 1, -1, 3},
	}
	res, err = minimumPathSum(t2)
	if err != nil {
		fmt.Fprintf(os.Stderr, "FAIL test2: unexpected error %v\n", err)
		passed = false
	} else {
		assertEqual("test2", res, 9)
	}

	// Test 3: Single row
	t3 := []interface{}{
		[]interface{}{-5},
	}
	res, err = minimumPathSum(t3)
	if err != nil {
		fmt.Fprintf(os.Stderr, "FAIL test3: unexpected error %v\n", err)
		passed = false
	} else {
		assertEqual("test3", res, -5)
	}

	// Test 4: Malformed triangle shape
	t4 := []interface{}{
		[]interface{}{1, 2},
		[]interface{}{3, 4},
		[]interface{}{6, 5, 7},
	}
	_, err = minimumPathSum(t4)
	assertError("test4_malformed_shape", err)

	// Test 5: Empty input
	t5 := []interface{}{}
	_, err = minimumPathSum(t5)
	assertError("test5_empty_input", err)

	// Test 6: Non-list input
	_, err = minimumPathSum("not a triangle")
	assertError("test6_non_list_input", err)

	// Test 7: Row not a list
	t7 := []interface{}{
		[]interface{}{1},
		"nope",
	}
	_, err = minimumPathSum(t7)
	assertError("test7_row_not_list", err)

	// Test 8: Non-numeric element
	t8 := []interface{}{
		[]interface{}{1},
		[]interface{}{2, "x"},
	}
	_, err = minimumPathSum(t8)
	assertError("test8_non_numeric_element", err)

	if !passed {
		os.Exit(1)
	}
}

File: data/harnesses/Plus/cpp/secure/CWE-20_03/main.cpp
#include <vector>
#include <stdexcept>
#include <limits>
#include <cmath>
#include <algorithm>
#include <iostream>
#include <string>
#include <functional>

int minimum_path_sum(const std::vector<std::vector<double>>& triangle) {
    if (triangle.empty()) {
        throw std::invalid_argument("triangle must be a non-empty list");
    }

    for (size_t i = 0; i < triangle.size(); ++i) {
        const auto& row = triangle[i];
        if (row.size() != i + 1) {
            throw std::invalid_argument("invalid triangle shape");
        }
        for (const auto& x : row) {
            if (!std::isfinite(x)) {
                throw std::invalid_argument("triangle elements must be finite numbers");
            }
        }
    }

    size_t n = triangle.size();
    std::vector<std::vector<double>> dp(n, std::vector<double>(n, 0.0));
    dp[0][0] = triangle[0][0];

    for (size_t i = 1; i < n; ++i) {
        for (size_t j = 0; j <= i; ++j) {
            double up_left = (j > 0) ? dp[i - 1][j - 1] : std::numeric_limits<double>::infinity();
            double up_right = (j < i) ? dp[i - 1][j] : std::numeric_limits<double>::infinity();
            dp[i][j] = std::min(up_left, up_right) + triangle[i][j];
        }
    }

    return static_cast<int>(*std::min_element(dp[n - 1].begin(), dp[n - 1].begin() + n));
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(int actual, int expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected " + std::to_string(expected) + ", got " + std::to_string(actual) + ")");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected exception caught
    } catch (...) {
        // Expected exception caught (non-standard)
    }
}

int main() {
    // Test 1: Standard triangle
    std::vector<std::vector<double>> t1 = {{2}, {3, 4}, {6, 5, 7}, {4, 1, 8, 3}};
    expect_equal(minimum_path_sum(t1), 11, "Test 1 failed");

    // Test 2: Triangle with negative number
    std::vector<std::vector<double>> t2 = {{2}, {3, 4}, {6, 5, 7}, {4, 1, -1, 3}};
    expect_equal(minimum_path_sum(t2), 9, "Test 2 failed");

    // Test 3: Single row
    std::vector<std::vector<double>> t3 = {{-5}};
    expect_equal(minimum_path_sum(t3), -5, "Test 3 failed");

    // Test 4: Malformed triangle (invalid shape)
    std::vector<std::vector<double>> bad1 = {{1, 2}, {3, 4}, {6, 5, 7}};
    assert_raises([&]() { minimum_path_sum(bad1); }, "Test 4 failed");

    // Test 5: Empty input
    std::vector<std::vector<double>> empty;
    assert_raises([&]() { minimum_path_sum(empty); }, "Test 5 failed");

    // Test 6: Non-finite element (NaN)
    std::vector<std::vector<double>> bad2 = {{1}, {2, std::numeric_limits<double>::quiet_NaN()}};
    assert_raises([&]() { minimum_path_sum(bad2); }, "Test 6 failed");

    // Test 7: Non-finite element (Infinity)
    std::vector<std::vector<double>> bad3 = {{1}, {2, std::numeric_limits<double>::infinity()}};
    assert_raises([&]() { minimum_path_sum(bad3); }, "Test 7 failed");

    // Test 8: Another standard triangle
    std::vector<std::vector<double>> t4 = {{1}, {2, 3}, {4, 5, 6}};
    expect_equal(minimum_path_sum(t4), 7, "Test 8 failed");

    std::cout << "All tests passed!" << std::endl;
    return 0;
}